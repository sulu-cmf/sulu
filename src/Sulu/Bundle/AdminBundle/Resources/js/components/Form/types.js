// @flow

export type FormFieldTypes = Array<{label: string, value: number}>;
