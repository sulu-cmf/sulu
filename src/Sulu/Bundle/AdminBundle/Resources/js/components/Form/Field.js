// @flow
import React from 'react';
import classNames from 'classnames';
import type {Node} from 'react';
import Grid from '../Grid';
import type {Size} from '../Grid';
import fieldStyles from './field.scss';
import gridStyles from './grid.scss';

type Props = {|
    children: Node,
    description?: string,
    error?: string,
    label?: string,
    required: boolean,
    size: Size,
    spaceAfter: Size,
|};

export default class Field extends React.Component<Props> {
    static defaultProps = {
        required: false,
        size: 12,
        spaceAfter: 0,
    };

    render() {
        const {children, description, error, label, required, size, spaceAfter} = this.props;

        const fieldClass = classNames(
            fieldStyles.field,
            {
                [fieldStyles.error]: !!error,
            }
        );

        return (
            <Grid.Item
                className={gridStyles.gridItem}
                size={size}
                spaceAfter={spaceAfter}
            >
                <div className={fieldClass}>
                    {label && <label className={fieldStyles.label}>{label}{required && ' *'}</label>}
                    {children}
                    {description &&
                    <label className={fieldStyles.descriptionLabel}>
                        {description}
                    </label>
                    }
                    <label className={fieldStyles.errorLabel}>
                        {error}
                    </label>
                </div>
            </Grid.Item>
        );
    }
}
