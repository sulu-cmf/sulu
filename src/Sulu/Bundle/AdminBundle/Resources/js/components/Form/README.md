The `Form` component allows to render fields in a structured form with labels, errors and so on. The form elements can
also be further structured by using `Section`s.

```javascript
const Form = require('./Form').default;

<Form>
    <Form.Section label="Section 1" size={3}>
        <Form.Field label="Author">
            <input type="text" style={{width: '100%'}} />
        </Form.Field>
    </Form.Section>
    <Form.Section label="Section 2" size={9}>
        <Form.Field description="This is the title" label="Title" required={true} size={3}>
            <input type="text" style={{width: '100%'}} />
        </Form.Field>
        <Form.Field description="This is the second title" error="Error!" label="Second title" required={true} size={9}>
            <input type="text" style={{width: '100%'}} />
        </Form.Field>
        <Form.Field description="Article" label="Article">
            <input type="text" style={{width: '100%'}} />
        </Form.Field>
    </Form.Section>
    <Form.Field label="Value">
        <input type="text" style={{width: '100%'}} />
    </Form.Field>
</Form>
```
