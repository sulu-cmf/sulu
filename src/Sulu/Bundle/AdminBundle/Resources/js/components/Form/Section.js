// @flow
import React from 'react';
import type {Node} from 'react';
import Divider from '../Divider';
import Grid from '../Grid';
import type {Size} from '../Grid';
import gridStyles from './grid.scss';

type Props = {|
    children: Node,
    label?: string,
    size: Size,
|};

export default class Section extends React.Component<Props> {
    static defaultProps = {
        size: 12,
    };

    render() {
        const {children, label, size} = this.props;

        return (
            <Grid.Section className={gridStyles.gridSection} size={size}>
                {label &&
                    <Grid.Item size={12}>
                        <Divider>
                            {label}
                        </Divider>
                    </Grid.Item>
                }
                {children}
            </Grid.Section>
        );
    }
}
