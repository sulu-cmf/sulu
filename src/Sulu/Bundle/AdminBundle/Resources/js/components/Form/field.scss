@import '../../containers/Application/colors.scss';
@import './variables.scss';

$labelColor: $scorpion;
$labelErrorColor: $roman;
$errorLabelColor: $persianRed;
$errorLabelBackgroundColor: $roman;

$errorLabelHeight: 20px;

.field {
    margin-bottom: calc($fieldMarginBottom - $errorLabelHeight);
}

.type {
    background: transparent;
    border: 0;
    color: $labelColor;
    cursor: pointer;
    margin-left: 7px;
    font-size: 12px;
}

.type-delimiter {
    margin-right: 7px;
}

.type-icon {
    margin-left: 3px;
    font-size: 10px;
}

.error {
    & > .label {
        color: $labelErrorColor;
    }
}

.label {
    display: block;
    color: $labelColor;
    font-size: 14px;
    line-height: 19px;
    margin-bottom: 10px;
    text-align: left;
    white-space: nowrap;

    &::after {
        content: '\200B';
    }
}

.description-label,
.error-label {
    display: block;
    color: $labelColor;
    font-size: 10px;
    line-height: $errorLabelHeight;
    height: $errorLabelHeight;
}

.error-label {
    color: $errorLabelColor;
}
