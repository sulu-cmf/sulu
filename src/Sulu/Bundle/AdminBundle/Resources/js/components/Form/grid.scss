@import '../../components/Grid/variables.scss';
@import '../../components/Form/variables.scss';
@import './variables.scss';
@import '../Divider/variables.scss';

.grid {
    margin-bottom: -$sectionMarginBottom;
    contain: layout;
}

.grid-section {
    margin-bottom: -$fieldMarginBottom;

    &:first-child {
        > .divider-container {
            margin-top: -$dividerTopMargin;
        }
    }

    &:last-child {
        margin-bottom: 0;
    }

    & + .grid-item {
        margin-top: $fieldMarginBottom;
    }
}

.grid-item + .grid-section {
    margin-top: -$fieldMarginBottom;
}

@media (max-width: $gridMinWidth) {
    .grid-section + .grid-section {
        margin-top: $fieldMarginBottom;
    }
}
