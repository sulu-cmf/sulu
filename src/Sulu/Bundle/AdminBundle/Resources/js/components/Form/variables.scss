$fieldMarginBottom: 30px;
$sectionMarginBottom: 30px;
