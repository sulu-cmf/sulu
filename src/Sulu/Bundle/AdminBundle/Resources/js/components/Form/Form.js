// @flow
import React from 'react';
import type {Node} from 'react';
import Grid from '../Grid';
import Field from './Field';
import Section from './Section';
import gridStyles from './grid.scss';

type Props = {|
    children: Node,
|};

export default class Form extends React.Component<Props> {
    static Field = Field;
    static Section = Section;

    render() {
        const {children} = this.props;

        return (
            <Grid className={gridStyles.grid}>
                {children}
            </Grid>
        );
    }
}
