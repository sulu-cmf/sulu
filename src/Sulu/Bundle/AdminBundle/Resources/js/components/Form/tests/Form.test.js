// @flow
import React from 'react';
import {render} from 'enzyme';
import Form from '../Form';

test('Render a form', () => {
    expect(render(<Form>Test</Form>)).toMatchSnapshot();
});
