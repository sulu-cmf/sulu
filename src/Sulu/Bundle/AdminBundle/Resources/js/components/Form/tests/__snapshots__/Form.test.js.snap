// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a form 1`] = `
<div
  class="grid grid"
>
  Test
</div>
`;
